src/vm/checker/typecheck/natives/options.rs

```diff
@@ -0,0 +1,90 @@
+    if args.len() != 1 {
+        return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(1, args.len())))        
+    }
+    
+    let inner_type = checker.type_check(&args[0], context)?;
+    let resp_type = TypeSignature::new_atom(
+        AtomTypeIdentifier::ResponseType(Box::new((no_type(), inner_type.clone()))));
+    Ok(resp_type)
+}
+
+pub fn check_special_is_okay(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    if args.len() != 1 {
+        return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(1, args.len())))
+    }
+    
+    let input = checker.type_check(&args[0], context)?;
+
+    if let Some(AtomTypeIdentifier::ResponseType(_types)) = input.match_atomic() {
+        return Ok(TypeSignature::new_atom(AtomTypeIdentifier::BoolType))
+    } else {
+        return Err(CheckError::new(CheckErrors::ExpectedResponseType))
+    }
+}
+
+pub fn check_special_is_none(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    if args.len() != 1 {
+        return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(1, args.len())))
+    }
+    
+    let input = checker.type_check(&args[0], context)?;
+
+    if let Some(AtomTypeIdentifier::OptionalType(_type)) = input.match_atomic() {
+        return Ok(TypeSignature::new_atom(AtomTypeIdentifier::BoolType))
+    } else {
+        return Err(CheckError::new(CheckErrors::ExpectedOptionalType))
+    }
+}
+
+pub fn check_special_default_to(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    if args.len() != 2 {
+        return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(2, args.len())))        
+    }
+    
+    let default = checker.type_check(&args[0], context)?;
+    let input = checker.type_check(&args[1], context)?;
+
+    if let Some(AtomTypeIdentifier::OptionalType(input_type)) = input.match_atomic() {
+        let contained_type = (**input_type).clone();
+        TypeSignature::most_admissive(default.clone(), contained_type.clone())
+            .ok_or(CheckError::new(CheckErrors::DefaultTypesMustMatch(contained_type, default)))
+    } else {
+        return Err(CheckError::new(CheckErrors::ExpectedOptionalType))
+    }
+}
+
+pub fn check_special_expects(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    if args.len() != 2 {
+        return Err(CheckError::new(CheckErrors::IncorrectArgumentCount(2, args.len())))        
+    }
+    
+    let input = checker.type_check(&args[0], context)?;
+    let on_error = checker.type_check(&args[1], context)?;
+
+    checker.track_return_type(on_error)?;
+
+    match input.match_atomic() {
+        Some(AtomTypeIdentifier::OptionalType(input_type)) => Ok((**input_type).clone()),
+        Some(AtomTypeIdentifier::ResponseType(response_type)) => Ok((**response_type).0.clone()),
```

Comment:
Not a bug, but if you find yourself using tuple indexes publicly (i.e. the `.0`) on 1-tuple items, one way around it is to implement the `Deref` and `DerefMut` traits for your struct and have the struct deref to its contained item.  Then, you won't need to do this.  Moreover, the generated code will be slightly more efficient -- the `Deref` and `DerefMut` traits allow the compiler to generate code that directly manipulates the contained object instead of dereferencing it via `.0`.

Reply:
Good to know -- not sure if it would help here though, since ResponseType is a 2-tuple.

Reply:
Ah lol saw that.